CONTRIBUTING.rst

```diff
@@ -178,9 +178,40 @@ not run on Travis jobs triggered by push requests. If your PR changes anything t
 or the tests themselves, Travis will detect this and run the nightly tests automatically with the
 regular build.
 
+If you need to retrigger Travis tests on a PR, you can push an empty commit to your branch. To create
+an empty commit, you can use the ``--allow-empty` option, e.g.
+``git commit --allow-empty -m "Trigger rebuild"``. Note that this will retrigger an entire rebuild -
+it is currently not possible to retrigger individual tests.
+
 If opening a PR that changes or adds new APIs, please update or add Python documentation as
 described in `Writing Docs`_ and commit the docs to your PR branch.
 
+Writing Python Tests
+--------------------
+If your PR includes code that isn't currently covered by our tests (e.g. adding a new flavor, adding
+autolog support to a flavor, etc.), you should write tests that cover your new code. MLflow currently
+uses ``pytest==3.2.1`` for testing. Your tests should be added to the relevant file under ``tests``, or
+if there is no appropriate file, in a new file prefixed with ``test_`` so that ``pytest`` includes that
+file for testing.
+
+If your tests require usage of a tracking URI, you can use the
+`pytest fixture <https://docs.pytest.org/en/3.2.1/fixture.html>`_
+`tracking_uri_mock <https://github.com/mlflow/mlflow/blob/master/tests/projects/utils.py#L74>`_
+to set up a mock tracking URI that will set itself up before your test runs and tear itself down after.
+
+If you are adding new framework flavor support, you'll need to modify ``pytest`` and Travis configurations so tests for your code can run properly. Generally, the files you'll have to edit are:
```

Comment:
maybe add a new subheading here. The following part talks about adding a new model flavor.